src/stream/src/executor/backfill/utils.rs

```diff
@@ -0,0 +1,477 @@
+                let order = cmp_datum_iter(lhs.iter(), rhs.iter(), pk_order.iter().copied());
+                match order {
+                    Ordering::Less | Ordering::Equal => true,
+                    Ordering::Greater => false,
+                }
+            }
+        };
+        new_visibility.append(v);
+    }
+    let (columns, _) = data.into_parts();
+    Ok(StreamChunk::new(
+        ops,
+        columns,
+        Some(new_visibility.finish()),
+    ))
+}
+
+/// Mark chunk:
+/// For each row of the chunk, forward it to downstream if its pk <= `current_pos`, otherwise
+/// ignore it. We implement it by changing the visibility bitmap.
+fn mark_chunk_inner(
+    chunk: StreamChunk,
+    current_pos: &OwnedRow,
+    pk_in_output_indices: PkIndicesRef<'_>,
+    pk_order: &[OrderType],
+) -> StreamChunk {
+    let (data, ops) = chunk.into_parts();
+    let mut new_visibility = BitmapBuilder::with_capacity(ops.len());
+    // Use project to avoid allocation.
+    for v in data.rows().map(|row| {
+        let lhs = row.project(pk_in_output_indices);
+        let rhs = current_pos.project(pk_in_output_indices);
+        let order = cmp_datum_iter(lhs.iter(), rhs.iter(), pk_order.iter().copied());
+        match order {
+            Ordering::Less | Ordering::Equal => true,
+            Ordering::Greater => false,
+        }
+    }) {
+        new_visibility.append(v);
+    }
+    let (columns, _) = data.into_parts();
+    StreamChunk::new(ops, columns, Some(new_visibility.finish()))
+}
+
+/// Builds a new stream chunk with `output_indices`.
+pub(crate) fn mapping_chunk(chunk: StreamChunk, output_indices: &[usize]) -> StreamChunk {
+    let (ops, columns, visibility) = chunk.into_inner();
+    let mapped_columns = output_indices.iter().map(|&i| columns[i].clone()).collect();
+    StreamChunk::new(ops, mapped_columns, visibility)
+}
+
+fn mapping_watermark(watermark: Watermark, upstream_indices: &[usize]) -> Option<Watermark> {
+    watermark.transform_with_indices(upstream_indices)
+}
+
+pub(crate) fn mapping_message(msg: Message, upstream_indices: &[usize]) -> Option<Message> {
+    match msg {
+        Message::Barrier(_) => Some(msg),
+        Message::Watermark(watermark) => {
+            mapping_watermark(watermark, upstream_indices).map(Message::Watermark)
+        }
+        Message::Chunk(chunk) => Some(Message::Chunk(mapping_chunk(chunk, upstream_indices))),
+    }
+}
+
+/// Gets progress per vnode, so we know which to backfill.
+pub(crate) async fn get_progress_per_vnode<S: StateStore, const IS_REPLICATED: bool>(
+    state_table: &StateTableInner<S, BasicSerde, IS_REPLICATED>,
+    state_len: usize,
+) -> StreamExecutorResult<Vec<(VirtualNode, BackfillProgressPerVnode)>> {
+    debug_assert!(!state_table.vnode_bitmap().is_empty());
+    let vnodes = state_table.vnodes().iter_vnodes();
+    let mut result = Vec::with_capacity(state_table.vnodes().len());
+    for vnode in vnodes {
```

Comment:
Since `get_progress_per_vnode` is called when we need to init the backfill state, cache of state table is highly likely to be empty and each `get_row` will be an actual S3 I/O. How about using `futures::future::try_join_all` to issue `state_table.get_row` concurrently to reduce overall latency?

Reply:
Updated https://github.com/risingwavelabs/risingwave/pull/10266/commits/9cc6ad26b19a67e4aa9c7e15246b0e885135f243